### ISSUE TITLE
`solana-cli::program` tests occasionally failing in CI

### ISSUE THREAD

**Issue by bw-solana** [solana-labs/solana#29688]
#### Problem
This suite of 9 tests experiences intermittent failures (~5%) due to failing to start the validator with the following panic message:

`panicked at 'validator start failed: "Address already in use (os error 98)"', test-validator/src/lib.rs:556:14`

After digging in further, here is an error message closer to the point of failure:
```
JSON RPC service unavailable error: Os { code: 98, kind: AddrInUse, message: "Address already in use" }.
Also, check that port 1092 is not already in use by another application
```
It appears there is some conflict with the JSON RPC socket address port we attempt to use.

[Example failing CI](https://buildkite.com/solana-labs/solana/builds/88085#0185a29d-1fc3-4391-9a8d-e9b2b802aff2)

#### Proposed Solution
TBD

**Comment by brennanwatt** [comment]
Looks like the RPC service gets assigned a random port between 1024 and 65536. I've seen this failure occur on the following ports:
- 1092
- 1115
- 4999

It doesn't look like any of the other RPC services running in parallel tests were using those ports, but it's possible some other service was. None of these ports were active before running the test

**Comment by brennanwatt** [comment]
Captured more details from an instance where ports from two concurrent tests collided:
```
ThreadId(10) new ContactInfo: LegacyContactInfo {
    id: EdEbRHcLF9fqNXZmwo23JpzGQy5Hv2TAfKocWF8E3fUH,
    gossip: 127.0.0.1:1093,
    tvu: 127.0.0.1:1094,
    tvu_forwards: 127.0.0.1:1095,
    repair: 127.0.0.1:1110,
    tpu: 127.0.0.1:1103,
    tpu_forwards: 127.0.0.1:1106,
    tpu_vote: 127.0.0.1:1107,
    rpc: 127.0.0.1:1115,
    rpc_pubsub: 127.0.0.1:26576,
    serve_repair: 127.0.0.1:1111,
    wallclock: 1673976827433,
    shred_version: 0,
}
ThreadId(7) new ContactInfo: LegacyContactInfo {
    id: 7oBG8UWXe3QDn2MoMwfxbu18WeW8e85ywNsXvkT2ZfGC,
    gossip: 127.0.0.1:1115,
    tvu: 127.0.0.1:1116,
    tvu_forwards: 127.0.0.1:1117,
    repair: 127.0.0.1:1122,
    tpu: 127.0.0.1:1118,
    tpu_forwards: 127.0.0.1:1119,
    tpu_vote: 127.0.0.1:1120,
    rpc: 127.0.0.1:11011,
    rpc_pubsub: 127.0.0.1:11263,
    serve_repair: 127.0.0.1:1123,
    wallclock: 1673976827466,
    shred_version: 0,
}
...
JSON RPC service unavailable error: Os { code: 98, kind: AddrInUse, message: "Address already in use" }.
Also, check that 0.0.0.0:1115 is not already in use by another application
```

Looks like the RPC port from thread 10 collided with the Gossip port from thread 7.

I would have expected the bind call to fail for addr in use (and presumably a different port to be selected), but that does not appear to be happening. More investigation needed to understand why

**Comment by brennanwatt** [comment]
Root cause: There is a race condition between (1) finding an available port for RPC and (2) actually using this port. When running concurrent testing, it is possible another thread sneaks in and binds this port between steps 1 and 2.

In `fn find_available_port_in_range`, we call `bind_common`, which calls `bind`, to determine the port is available. However, the UDP socket is dropped after determining the port is available. Upon being dropped, the port is now up for grabs again and can get sniped before we actually use the port later when setting up the JSON RPC service.

Note: I don't believe we have any major problem outside of testing scenarios because we check for RPC ports last (i.e. after Gossip, TVU, TPU, etc.) and so nothing is expected to attempt to snipe our ports. That said, some outside application could come in and try and reserve the ports.

**Comment by brennanwatt** [comment]
Created draft PRs for two possible solutions:
- #29843 - This binds to the RPC port as a means of reserving it and then releases it immediately before starting JSON RPC service. This technically doesn't completely close the timing hole, but it shrinks it drastically
- #29844 - This uses a shared, static mutex between tests to serialize the TestValidator start process. This could mean slowing down testing a little bit by forcing serialization and doesn't prevent outside applications from sniping the port (though that's probably outside the scope of this issue)

**Comment by brennanwatt** [comment]
Somehow still seeing a port collision in [CI](https://buildkite.com/solana-labs/solana/builds/88729#0185e457-8f28-4b9c-a88d-c90fb1ba0cce) for #29844 . Not yet clear to me how this might be happening

**Comment by jbiseda** [comment]
> Somehow still seeing a port collision in [CI](https://buildkite.com/solana-labs/solana/builds/88729#0185e457-8f28-4b9c-a88d-c90fb1ba0cce) for #29844 . Not yet clear to me how this might be happening

Could the pubsub socket be in `TIME_WAIT` state? `SO_REUSEADDR`/`SO_REUSEPORT` may be helpful.

**Comment by bw-solana** [comment]
> Could the pubsub socket be in `TIME_WAIT` state? `SO_REUSEADDR`/`SO_REUSEPORT` may be helpful.

Seems possible... I'll give this a go.

It looks like the JSON RPC server crate we're using only allows for port reuse when multiple threads are being spun up, but we can cross that bridge once the root cause is confirmed

**Comment by bw-solana** [comment]
@jbiseda - Update on this: I suspect you are right, but can't say for sure.

Running in a loop, I'm seeing failure ~1/10k iterations. Either of the following mitigations appear to have resolved the issue (not seeing failure after 100k iterations):

1. launch RPC server w/ 2 threads (hack to enable port reuse)
2. If server fails to start, sleep for 30 seconds and retry (repeat this up to 10 times)

For 1, I'm worried we could accidentally reuse the port in cases where we don't want to and trade a relatively small and easy problem for a harder one. Not to mention we would want to change the `jsonrpc-http-server` crate.

Maybe 2 is worth pursuing.. failure rate is low enough we could probably get away with just ignoring. Also, CI might still fail with a timeout if we have to wait a long time for `TIME_WAIT` state transition (could be minutes).

**Comment by jbiseda** [comment]
> For 1, I'm worried we could accidentally reuse the port in cases where we don't want to and trade a relatively small and easy problem for a harder one. Not to mention we would want to change the `jsonrpc-http-server` crate.

I agree that it's probably not worth adding this complexity to work around the problem.

> Maybe 2 is worth pursuing.. failure rate is low enough we could probably get away with just ignoring. Also, CI might still fail with a timeout if we have to wait a long time for `TIME_WAIT` state transition (could be minutes).

I think the TIME_WAIT timeout is 180s. But if the retry selects a different random port this would probably succeed with successive retries. This may be a simple way to reduce the failures.

**Comment by bw-solana** [comment]
> I think the TIME_WAIT timeout is 180s. But if the retry selects a different random port this would probably succeed with successive retries. This may be a simple way to reduce the failures.

This would be nice except that we've already setup a bunch of structs with the RPC contact info before launching the RPC server. It's also why the pre-ordainment of port number happens so early and creates a wide timing hole for something else to come and snipe the port

**Comment by bw-solana** [comment]
Closing this issue as low priority. It's a test only issue (or at least doesn't happen during normal runtime) and infrequent enough that we can live with it
